mountpoint-s3/src/fs.rs

```diff
@@ -411,6 +414,30 @@ impl S3Personality {
     }
 }
 
+/// Server-side encryption configuration for newly created objects
+#[derive(Debug, Clone, Default)]
+pub struct ServerSideEncryption {
+    sse_type: Option<String>,
+    sse_kms_key_id: Option<String>,
+}
+
+impl ServerSideEncryption {
+    /// Construct settings from the raw values provided via CLI
+    pub fn new(sse_type: Option<String>, sse_kms_key_id: Option<String>) -> Self {
+        // TODO: compute checksum
+        Self {
+            sse_type,
+            sse_kms_key_id,
+        }
+    }
+
+    /// Verfify the checksum and return a the SSE type and the AWS KMS Key ID if provided
+    pub fn into_inner(&self) -> (Option<String>, Option<String>) {
+        // TODO: verify checksum
+        (self.sse_type.clone(), self.sse_kms_key_id.clone())
+    }
```

Comment:
A good reason to implement an `into_...` method and consume `self` is that we want to validate the content and return it directly, that is avoid any `clone` **after** verifying the checksum.

However, since this method is only useful once the checksum verification is implemented, we could also discuss it in that future PR.

Reply:
I thought about adding a enum here but after all it just looked redundant to me (correct me if I am wrong). I see the interface of this struct something similar to what we have in current revision, ie receiving strings (provided as CLI args) and returning strings to construct a `PutObject` request. That works well for KMS/KMS-DSSE, because we do not do anything different depending on KMS/KMS-DSSE/default, so we do not need pattern matching on enum variants.

Now SSE-C. Its true that it won't fit very well in current version of this struct. `sse_kms_key_id` argument of `new` will need to be something like `key_material` or `key_material_path_on_disk`. We will probably need to do some pattern matching depending on `sse_type` **inside** implementation of this struct, e.g. to read the material from disk in case of SSE-C. But still it seems like it better to be encapsulated by this struct, right? Meaning the interface will stay the same: `new` and `into_inner`. Thus even if we will want to add a enum in future, it will be a very straightforward change, affecting these 20 lines of `ServerSideEncryption` implementation only.

Reply:
I struggle to see how you would implement checksum verification in the `unpack` method (who copies the Strings and when matters).

May I suggest reverting to the two getter methods in this PR? We can think about the best approach when we add the checksums.

Reply:
Yep, let's just use two getters and tackle the checksumming in its own PR.

Reply:
yeah, a little of back and forth, but two getters seem better then returning a tuple